content/doc/tutorials/building-a-python-app-with-pyinstaller.adoc

```diff
@@ -0,0 +1,486 @@
+  that Jenkins uses to run the `Deliver` stage of your Pipeline project. This
+  container's lifespan lasts the duration of your `Deliver` stage's execution.
+<2> This `sh` step (of the link:/doc/book/pipeline/syntax/#steps[`steps`]
+section) executes the `pyinstaller` command (in the PyInstaller container) on
+your simple Python application. This bundles your `add2vals.py` Python
+application into a single standalone executable file (via the `--onefile`
+option) and outputs the this file to the `dist` workspace directory (within the
+Jenkins home directory). Although this step consists of a single command, as a
+general principle, it's a good idea to keep your Pipeline code (i.e. the
+`Jenkinsfile`) as tidy as possible and place more complex build steps
+(particularly for stages consisting of 2 or more steps) into separate shell
+script files like the `deliver.sh` file. This ultimately makes maintaining your
+Pipeline code easier, especially if your Pipeline gains more complexity.
+. Save your amended `Jenkinsfile` and commit it to your local
+  `simple-python-pyinstaller-app` Git repository. E.g. Within the
+  `simple-python-pyinstaller-app` directory, run the commands: +
+  `git stage .` +
+  then +
+  `git commit -m "Add 'Deliver' stage"`
+. Go back to Jenkins again, log in again if necessary and ensure you've accessed
+  Jenkins's Blue Ocean interface.
+. Click *Run* at the top left, then quickly click the *OPEN* link which appears
+  briefly at the lower-right to see Jenkins running your amended Pipeline
+  project. If you weren't able to click the **OPEN** link, click the _top_ row
+  on the Blue Ocean interface to access this feature. +
+  **Note:** It may take a few minutes for the `cdrx/pyinstaller-linux` Docker
+  image to download (if this hasn't already been done). +
+  If your amended Pipeline ran successfully, here's what the Blue Ocean
+  interface should look like. Notice the additional "Deliver" stage. Click on
+  the previous "Test" and "Build" stage circles to access the outputs from those
+  stages.
++
+[.boxshadow]
+image:tutorials/python-pyinstaller-21-deliver-stage-runs-successfully.png[alt="Deliver
+stage runs successfully",width=100%]
+
++
+Here's what the output of the "Deliver" stage should look like, showing you the
+results of PyInstaller bundling your Python application into a single standalone
+executable file.
+[.boxshadow]
+image:tutorials/python-pyinstaller-22-deliver-stage-output-only.png[alt="Deliver
+stage output only",width=100%]
+. Click the *X* at the top-right to return to the main Blue Ocean interface,
+  which lists your previous Pipeline runs in reverse chronological order.
+[.boxshadow]
+image:tutorials/python-pyinstaller-23-main-blue-ocean-interface-with-all-previous-runs-displayed.png[alt="Main
+Blue Ocean interface with all previous runs displayed",width=100%]
+
+
+=== Follow up (optional)
+
+If you use Linux, you can try running the standalone `add2vals` application you
```

Comment:
This is not a _great_ pattern to encourage the reader to perform. What should be done instead is the suggestion of adding an `archiveArtifacts` step to the "Deliver" stage, which would allow the user to download the artifact through the web UI and execute it.

Reply:
That's what I started thinking about late last night - I will update!

Reply:
All done!